etc/vf/vf.man
.TH VF LOCAL uw
. \" @(#)vfs.l	1.1 1/11/84
.UC 4
.SH NAME
VF - Verify File Owners, Groups, Modes, and Types
.SH SYNOPSIS
.I vf 
[-c] [file]
.SH DESCRIPTION
.I Vf
will consult the named file, or 
.I stdin 
if no file
is given, for file specifications. 
If there are any discrepancies between the
actual attributes of the named files and the way
these are specified, 
.I vf
will output an error message to this effect.
.PP
If the 
.I \-c
option is given, then
.I vf
will attempt to correct the attributes of the files.
Of course, some problems (file non-existent, file is 
a symbolic link) require human intervention.
.PP 
Bearing in mind that directory specifcations
.I must
have a trailing slash, as in 
.I /tmp/,
the input file should contain lines 
in the following 
form:
.TP
.I pathname owner group mode
.PP
Where 
.I owner
and 
.I group
are symbolic names and 
.I mode
is an octal protection code.
.SH DIAGNOSTICS
.I Vf
will attempt to process the whole list, even if 
simple errors, such as missing files, should occur during processing.  The
following exit stati are defined for the
.I vf
program:
.TP
.B EX_OK (0) normal exit
No inconsistencies were discovered between the input file's idea
of how things should be and the actual situation.
.TP
.B EX_USAGE (64) command line usage error
The
.I vf
program was called with an improper argument count.
.TP
.B EX_TEMPFAIL (75) temporary failure
All permission inconsistencies have been corrected by the program.
If rerun, 
.B vf
should exit EX_OK.
.TP 
.B EX_DATAERR (65) data format error
The input file contained inproperly formed input.
.B Vf
will notify the user of the line number in the input file
at which the error occurred.
.TP 
.B EX_UNAVAILABLE (69) service unavailable
Errors were encountered which could not be 
corrected by 
.B vf,
such as incorrect file types or missing files.  Divine
intervention (i.e. superuser) is required.
.TP
.B EX_NOPERM  (77) permission denied
The program was run with the
.B \-c
option, and only the superuser can correct file owners.
.TP
.B EX_NOINPUT (66) cannot open input
The specified input file did not exist.
.TP
.B EX_SOFTWARE (70) internal software error
The 
.I vf
program needs to be recompiled with a higher
constant array size used for storing user and group
names and numeric ids.
.SH FIXFILES
The 
.B fixfiles
script runs 
.B vf
every morning, collecting diagnostics and mailing
them to the system administrator.  No mail is
generated if 
.B vf
exits with no errors (EX_OK).  
.PP
.B Input
is set to /etc/files.perms
and a copy of the error messages is kept in 
/etc/files.errs,
in addition to being sent to the system adminsistrator (root).
.SH FUTURE WORK
There is not yet a facility for specifying the way 
files should be linked together.  
.SH BUGS
None known.
.SH AUTHOR
Tom Christiansen 


sys/CHANGES
************************ SYS ****************************************
init_main.c:
	added kmem_init, vnodes for swap and arg and initialized u.u_cred
init_sysent.c:
	added vfs, nfs, shmem and idm system calls
kern_acct.c:
	vnode changes
kern_clock.c:
	added routine uniqtime()
kern_descrip.c:
	vnode changes and cred routines
kern_exec.c:
	vnode changes, pgthresh, dnlc instead of nch
kern_exit.c:
	vnode changes, free credit, added check for valid mbuf in exit
kern_fork.c:
	vnode changes, added variable maxuprc to fork1
kern_mmap.c:
	include line changes only
kern_proc.c:
	include line changes only
kern_prot.c:
	cred changes and cred routines.
kern_resource.s:
	include line changes only
kern_sig.c:
	changed core() to use vnodes
kern_subr.c:
	include line changes only
kern_synch.c:
	include line changes only
kern_time.c:
	include line changes only
kern_xxx.c:
	added get/set domainname syscalls and vnode changes
subr_log.c:
	include line changes only
subr_mcount.c:
	no changes
subr_prf.c:
	include line changes only
subr_rmap.c:
	include line changes only
subr_xxx.c:
	include line changes only
sys_generic.c:
	include line changes only
sys_process.c:
	vnode changes
sys_socket.c:
	include line changes only
syscalls.c:
	added vfs syscalls to list
tty.c:
	include line changes and usestat mods
tty_bk.c:
	include line changes only
tty_conf.c:
	include line changes only
tty_pty.c:
	include line changes, O_NDELAY, and left BLIT mods because I think
	they are in under a different guise
tty_subr.c:
	no changes
tty_tb.c:
	include line changes only
tty_tty.c:
	include line changes only
uipc_domain.c:
	include line changes only
uipc_mbuf.c:
	added numberous mbuf routines
uipc_proto.c:
	no changes
uipc_socket.c:
	include lines, and a check for a null mbuf pointer
uipc_socket2.c:
	include line changes only
uipc_syscalls.c:
	added crfree 
uipc_usrreq.c:
	vnode changes
vfs.c:
	changed to 4.3 syscall style, changed UIOSEG_USER to UIO_USERSPACE
vfs_bio.c:
	include line changes only
vfs_conf.c:
	include line changes only
vfs_dev.c:
	include line changes only
vfs_dnlc.c:
	include line changes only
vfs_io.c:
	include line changes only
vfs_lookup.c:
	include line changes, UIO flags
vfs_pathname.c:
	include line changes, UIO flags
vfs_syscalls.c:
	changed to 4.3 syscall style
vfs_vnode.c:
	include line changes only
vfs_xxx.c:
	changed to 4.3 syscall style
vm_drum.c:
	vnode changes (swapdev --> swapdev_vp)
vm_mem.c:
	vnode changes, mfind, mhash, munhash and added mpurgevfs
vm_mon.c:
	no changes
vm_page.c:
	vnode changes, calls to mfind ...
vm_proc.c:
	include line changes only
vm_pt.c:
	vnode changes, calls to swap ...
vm_sched.c:
	vnode changes
vm_subr.c:
	vnode changes
vm_sw.c:
	vnode changes
vm_swap.c:
	change swapdev to swapdev_vp
vm_swp.c:
	vnode changes
vm_text.c:
	vnode changes for text cacheing
********************************* H ******************************
acct.h:
	changed acctp to vnode pointer
bk.h:
	unchanged
buf.h:
	added b_vp and B_NOCACHE
callout.h:
	unchanged
clist.h:
	unchanged
cmap.h:
	added c_vp to cmap and changed CMAPSZ from 16 to 20 to reflect addition
	of vp
conf.h:
	nochange
des.h:
	Sun data encryption standard driver (UNUSED)
dir.h:
	sun version of dir.h old dir.h moved to ufs/fsdir.h
dk.h:
	no change
dkbad.h:
	no change
dmap.h:
	no change
dnlc.h:
	Sun include file for directory name lookup cache
domain.h:
	no change
errno.h:
	added ESTALE and EREMOTE
exec.h:
	no change
file.h:
	added cred field, changed inode to vnode in comments, DTYPE_INODE to VNODE 
gprof.h:
	no change
ioctl.h:
	added usestat info, vicom ioctls, load avg and time ioctls
kernel.h:
	added domainname and domainnamelen (used by nfs);
map.h:
	no change
mbuf.h:
	added support for more types of mbufs specifically type 2 which specify
	their own free routine
mman.h:
	no change
mount.h:
	copy of sun h/mount.h old h/mount.h moved to ufs/mount.h
msgbuf.h:
	no change
mtio.h:
	no change
param.h:
	added a priority for VFS
pathname.h:
	copy of Sun include file
proc.h:
	added SYSCALLMETER stuff
protosw.h:
	no change
reboot.h:
	no change
resource.h:
	no change
seg.h:
	no change
signal.h:
	no change
socket.h:
	added AF_CRY
socketvar.h:
	no change
stat.h:
	no change
systm.h:
	added argdev_vp and swapdev_vp 
	added devtovp()
text.h
	changed inode pointer to vnode pointer, changed XPAGI to XPAGV
time.h
	no change
timeb.h:
	no change
times.h:
	no change
trace.h:
	no change
tty.h:
	added usestat code and TS_ONDELAY
ttychars.h:
	added usestat code
ttydev.h:
	no change
types.h:
	no change
uio.h:
	no change
un.h:
	no change
unpcb.h:
	changed reference from inode to vnode
user.h:
	changed to vnode references added ucred struct and references
	removed namei fields
vadvise.h:
	no change
vcmd.h:
	no change
vfs.h:
	copy of sun file
vlimit.h:
	no change
vm.h:
	no change
vmmac.h:
	no change
vmmeter.h:
	no change
vmparam.h:
	no change
vmsystm.h:
	no change
vnode.h:
	copy of sun file
vtimes.h:
	no change
wait.h:
	no change
**************************** UFS *********************************
*********** using a copy of sun ufs directory for now ************
fs.h:
	added id field
fsdir.h:
	no change
inode.h:
	lots of changes for vnodes
mount.h:
	lots of vnode and sun quota changes
quota.c:
	copy of sun file
quota.h:
	copy of sun file
quota_syscalls.c:
	copy of sun file
quota_ufs.c:
	copy of sun file
ufs_alloc.c:
	all calls to mfind, munhash, bread, and getblk were changed for vnodes
ufs_bmap.c:
	added sync option to bmap
ufs_dir.c:
	copy of sun file
ufs_dsort.c:
	no change
**************************** vax *******************************************
Locore.c:
	no change
autoconf.c:
	no change
vm_machdep.c:
	changed call to munhash


etc/vf/vf.c
/*
 *  vf.c by tom christiansen, Friday, 18 Jul 86 -- 06:56:22 CDT
 *
 *  read in list of files with attributes, compare with
 *  existing files' attrs, correct them if -c flag given,
 *  print diagnostics explaining discrepancies and corrections.
 *  
 * 	modified by author Wednesday, 30 Jul 86 -- 05:20:02 CDT 
 * 		to send all msgs to stdout
 * 		have exit status better reflect error conditions
 */
#include <stdio.h>
#include <sysexits.h>
#include <pwd.h>
#include <grp.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/file.h>
#include <sys/stat.h>

#ifndef MAXUSERS
#	define MAXUSERS	1500
#endif !MAXUSER

#define MODE(x) ((x) & ~S_IFMT) 
#define TYPE(x) ( ((x) & S_IFMT) == 0 ? S_IFREG : ((x) & S_IFMT))

#define FLAGS(A,B) (L_/**/A | L_/**/B)
#define L_USER		01
#define L_GROUP		02
#define L_NAME		10
#define L_ID		20

#define NOBODYS_UID	32767

#define STREQ(STR1,STR2) (strcmp(STR1,STR2) == 0)

extern char *malloc();
extern int   sys_nerr;
extern char *sys_errlist[];
extern int   errno;

char name[1025], owner[9], group[9];
char *users[MAXUSERS], *groups[MAXUSERS];
int errors=0, corrected=0, linecount=0;
struct stat stat_buf;
char *typename();

/*
 *  convenient lie:  assumes a (struct passwd *)
 *  and a (struct group *) both start like this.
 */
struct info {
	char *myname;
	char *mypasswd;
	int   myid;
};




main (ac,av) 
	char *av[];
{
	FILE *input;
	char *program = *av;
	int mode,cflag=0;
	register scan_count,uid,gid;
	register struct stat *sb = &stat_buf;

	if (STREQ(av[1],"-c")) {
		cflag++;
		av++;
		ac--;
	}

	if (cflag && getuid()) {
		fprintf ( stdout, "%s: Not superuser\n", program );
		exit (EX_NOPERM);
	}

	/*
	 *  process arguments
	 */
	switch (ac) {
		case 1:
			input = stdin;
			break;

		case 2:
			if (av[1][0] == '-')
				goto usage;
			if (!(input=fopen(av[1],"r"))) {
				fprintf ( stdout, "%s: ", program);
				perror (av[1]);
				exit (EX_NOINPUT);
			}
			break;

		default:
		usage:
			fprintf (stdout, "usage: %s [-c] [file]\n", program);
			exit (EX_USAGE);
	}

	/*
	 *  initialize all elts of the user and group arrays
	 *  to be "<unknown ....>".  we will check for 
	 *  whether (*users[uid] == '<') to see whether we
	 *  can chown it.
	 */
	init_array(users,&users[MAXUSERS],"<unknown user>");
	init_array(groups,&groups[MAXUSERS],"<unknown group>");

	/*
	 *  process input
	 */
	while(EOF!=(scan_count=fscanf(input,"%1024s%8s%8s%o",name,owner,group,&mode))) {
		linecount++;
		readln(input);	/* discard up to the next \n or EOF */

		if (scan_count != 4) {
			errors++;
			fprintf (stdout, "%s: format error at line %d in file \"%s\"\n", 
				program, linecount, (input == stdin) ? "<stdin>" : av[1] );
			fprintf (stdout,
						"\tscanned only %d items instead of 4\n",scan_count);
			exit(EX_DATAERR);
		}

		/*
		 *  is there such a file?
		 */
		if ( access (name, F_OK) != 0 || lstat (name, sb) < 0) {
			errors++;
			perror(name);
			continue;
		}

		/*
		 *  fix mode in case of directory specification
		 */
		{
			register char * s = name;
			while (s[1])
				s++;
			if (*s == '/')
				mode |= S_IFDIR;
		}

		/*
		 *  find numeric values of desired owner and group
		 */
		uid = getid ( owner, users, getpwnam, FLAGS(NAME,USER));
		gid = getid ( group, groups, getgrnam, FLAGS(NAME,GROUP));

		if (sb->st_uid != uid) {
			getid ( sb->st_uid, users, getpwuid, FLAGS(ID,USER));
			fprintf ( stdout,"%s: owner is \"%s\", expected \"%s\"\n", 
				name, users[sb->st_uid], owner);
		}
		if (sb->st_gid != gid)  {
			getid ( sb->st_gid, groups,  getgrgid, FLAGS(ID,GROUP));
			fprintf ( stdout, "%s: group is \"%s\", expected \"%s\"\n", 
				name, groups[sb->st_gid], group);
		}

		if (sb->st_uid != uid || sb->st_gid != gid)  {
			if (cflag) {
				if ( (uid != NOBODYS_UID && *users[uid] == '<') 
						|| *groups[gid] == '<' ) {
					errors++;	
					fprintf (stdout, "\tCANNOT CORRECT %s\n",name);
				}
				else {
					fprintf ( stdout, "   correcting owner.group to be \"%s.%s\"\n",
						(uid==NOBODYS_UID)?"nobody":users[uid], groups[gid] );
					if (chown ( name, uid, gid ) < 0) {
						errors++;
						perror("chown");
					} else
					corrected++;
				}
			} else
				errors++;
		}

		if (MODE(sb->st_mode) != MODE(mode))  {
			fprintf ( stdout, "%s: mode is %04o, expected %04o\n",
				name, MODE(sb->st_mode), MODE(mode));
			if (cflag) {
				fprintf ( stdout, "    correcting mode %04o to be %04o\n",
					MODE(sb->st_mode), MODE(mode));
				if (chmod ( name, MODE(mode)) < 0) {
					errors++;
					perror("chmod");
				} else
					corrected++;
			} else
				errors++;
		}
		if (TYPE(sb->st_mode) != TYPE(mode))  {
			errors++;
			fprintf ( stdout, "%s: type is \"%s\", expected \"%s\"\n",
				name, typename((int)sb->st_mode), typename(mode));
			fprintf (stdout, "\tCANNOT CORRECT %s\n",name);
		}

	}

	if (errors > 0)
		exit(EX_UNAVAILABLE);
	else if (corrected > 0)
		exit(EX_TEMPFAIL);
	else
		exit(EX_OK);
}



getid ( who, list, lookup, code )
	char *who;
	char **list;
	struct info * (*lookup)();  /* culpa mea: this is a convenient lie */
{
	register struct info *p;

	if ( ((code&L_NAME) && STREQ(who,"nobody")) || ((code&L_ID) && (int) who == NOBODYS_UID))
		return NOBODYS_UID;

	if (!(p = (*lookup)(who))) {
		errors++;
		fprintf (stdout, "%s: No such %s: ", name, (code&L_USER) ? "user" : "group");
		fprintf (stdout, (code&L_ID)? "%d\n"  : "%s\n", who );
		return -1;
	}

	if ( p->myid >= MAXUSERS ) {
		errors++;
		fprintf (stdout, "%s: TO STORE %s, RECOMPILE WITH MAXUSERS > %d\n",
			name,who,p->myid);
		exit (EX_SOFTWARE);
	}
	if ( list [p->myid] [0] == '<' )
		(void) strcpy ( list [ p->myid ] = 
							malloc ( (u_int) strlen (p->myname)+1), 
						p->myname );
	return p->myid ;
}

char *
typename(mode) {
	switch ( TYPE(mode) ) {
		case S_IFREG:
			return "plain file";
		case S_IFDIR: 
			return "directory";
		case S_IFCHR:
			return "character special";
		case S_IFBLK:
			return "block special";
		case S_IFLNK:
			return "symbolic link";
		case S_IFSOCK:
			return "socket";
		default:
			errors++;
			fprintf (stdout, "typename: can't match mode %08o (%08o)\n",
				TYPE(mode),mode);
			return "unknown file type";
	}
}

init_array ( start, end, string )
	register char **start,**end,*string;
{
	do {
		*start++ = string;
	} while (start <= end);
}



readln(file)
	FILE *file;
{
	int wasted;

	while ( (wasted = getc(file)) != EOF && wasted != '\n')
		/* VOID */;
}

/*
 *  perror() to stdout
 */
perror(s)
	char *s;
{
	int oerrno = errno;
	fprintf (	stdout, "%s%s%s\n", s,
				( s != NULL )
					? ": " 
					: "" , 
				(oerrno < sys_nerr)
					? sys_errlist[oerrno]
					: "Unknown error" 
			);
}

